File: slang-checks/src/main/java/org/sonarsource/slang/checks/HardcodedCredentialsCheck.java
/*
 * SonarSource SLang
 * Copyright (C) 2018-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonarsource.slang.checks;

import java.net.URI;
import java.net.URISyntaxException;
import org.sonarsource.slang.api.AssignmentExpressionTree;
import org.sonarsource.slang.api.StringLiteralTree;
import org.sonarsource.slang.api.Tree;
import org.sonarsource.slang.api.VariableDeclarationTree;
import org.sonarsource.slang.checks.api.CheckContext;
import org.sonarsource.slang.checks.api.InitContext;
import org.sonarsource.slang.checks.api.SlangCheck;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.sonar.check.Rule;
import org.sonar.check.RuleProperty;
import org.sonarsource.slang.checks.utils.ExpressionUtils;

import javax.annotation.Nullable;

@Rule(key = "S2068")
public class HardcodedCredentialsCheck implements SlangCheck {

  private static final String DEFAULT_VALUE = "password,passwd,pwd,passphrase";
  private static final Pattern URI_PREFIX = Pattern.compile("^\\w{1,8}://");

  @RuleProperty(
    key = "credentialWords",
    description = "Comma separated list of words identifying potential credentials",
    defaultValue = DEFAULT_VALUE)
  public String credentialWords = DEFAULT_VALUE;

  private List<Pattern> variablePatterns;
  private List<Pattern> literalPatterns;

  @Override
  public void initialize(InitContext init) {
    init.register(AssignmentExpressionTree.class, (ctx, tree) -> {
      Tree leftHandSide = tree.leftHandSide();
      ExpressionUtils.getMemberSelectOrIdentifierName(leftHandSide)
          .ifPresent(variableName -> checkVariable(ctx, leftHandSide, variableName, tree.statementOrExpression()));
    });

    init.register(VariableDeclarationTree.class, (ctx, tree) ->
      checkVariable(ctx, tree.identifier(), tree.identifier().name(), tree.initializer())
    );

    init.register(StringLiteralTree.class, (ctx, tree) -> {
      String content = tree.content();
      if (isURIWithCredentials(content)) {
        ctx.reportIssue(tree, "Review this hard-coded URL, which may contain a credential.");
      } else {
        literalPatterns()
          .map(pattern -> pattern.matcher(content))
          .filter(Matcher::find)
          .map(matcher -> matcher.group(1))
          .filter(match -> !isQuery(content, match))
          .forEach(credential -> report(ctx, tree, credential));
      }
    });
  }

  private static boolean isURIWithCredentials(String stringLiteral) {
    if (URI_PREFIX.matcher(stringLiteral).find()) {
      try {
        String userInfo = new URI(stringLiteral).getUserInfo();
        if (userInfo != null) {
          String[] parts = userInfo.split(":");
          return parts.length > 1 && !parts[0].equals(parts[1]);
        }
      } catch (URISyntaxException e) {
        // ignore, stringLiteral is not a valid URI
      }
    }
    return false;
  }

  private static boolean isNotEmptyString(@Nullable Tree tree) {
    return tree instanceof StringLiteralTree
      && !((StringLiteralTree)tree).content().isEmpty();
  }

  private static boolean isQuery(String value, String match) {
    String followingString = value.substring(value.indexOf(match) + match.length());
    return followingString.startsWith("=?")
      || followingString.startsWith("=%")
      || followingString.startsWith("=:")
      || followingString.startsWith("={") // string format
      || followingString.equals("='");
  }

  private static void report(CheckContext ctx, Tree tree, String matchName) {
    String message = String.format("\"%s\" detected here, make sure this is not a hard-coded credential.", matchName);
    ctx.reportIssue(tree, message);
  }

  private void checkVariable(CheckContext ctx, Tree variable, String variableName, @Nullable Tree value) {
    if (isNotEmptyString(value)) {
      variablePatterns()
        .map(pattern -> pattern.matcher(variableName))
        .filter(Matcher::find)
        .forEach(matcher -> checkAssignedValue(ctx, matcher, variable, ((StringLiteralTree) value).value()));
    }
  }

  private static void checkAssignedValue(CheckContext ctx, Matcher matcher, Tree leftHand, String value) {
    if (!matcher.pattern().matcher(value).find()) {
      report(ctx, leftHand, matcher.group(1));
    }
  }

  private Stream<Pattern> variablePatterns() {
    if (variablePatterns == null) {
      variablePatterns = toPatterns("");
    }
    return variablePatterns.stream();
  }

  private Stream<Pattern> literalPatterns() {
    if (literalPatterns == null) {
      literalPatterns = toPatterns("=\\S");
    }
    return literalPatterns.stream();
  }

  private List<Pattern> toPatterns(String suffix) {
    return Stream.of(credentialWords.split(","))
      .map(String::trim)
      .map(word -> Pattern.compile("(" + word + ")" + suffix, Pattern.CASE_INSENSITIVE))
      .collect(Collectors.toList());
  }

}

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/HardcodedCredentialsCheck.kt
/*
 * SonarSource SLang
 * Copyright (C) 2018-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonarsource.kotlin.checks

import java.net.URI
import java.net.URISyntaxException
import org.jetbrains.kotlin.com.intellij.psi.PsiElement
import org.jetbrains.kotlin.lexer.KtTokens
import org.jetbrains.kotlin.psi.KtBinaryExpression
import org.jetbrains.kotlin.psi.KtDotQualifiedExpression
import org.jetbrains.kotlin.psi.KtElement
import org.jetbrains.kotlin.psi.KtExpression
import org.jetbrains.kotlin.psi.KtNameReferenceExpression
import org.jetbrains.kotlin.psi.KtProperty
import org.jetbrains.kotlin.psi.KtStringTemplateExpression
import org.sonar.check.Rule
import org.sonar.check.RuleProperty
import org.sonarsource.kotlin.api.AbstractCheck
import org.sonarsource.kotlin.plugin.KotlinFileContext

/**
 * Replacement for [org.sonarsource.slang.checks.HardcodedCredentialsCheck]
 */
@Rule(key = "S2068")
class HardcodedCredentialsCheck : AbstractCheck() {
    @RuleProperty(key = "credentialWords",
        description = "Comma separated list of words identifying potential credentials",
        defaultValue = DEFAULT_VALUE)
    var credentialWords = DEFAULT_VALUE
    private var variablePatterns: Sequence<Regex>? = null
    private var literalPatterns: Sequence<Regex>? = null

    companion object {
        private const val DEFAULT_VALUE = "password,passwd,pwd,passphrase"
        private val URI_PREFIX = Regex("^\\w{1,8}://")

        private fun isURIWithCredentials(stringLiteral: String): Boolean {
            if (URI_PREFIX.containsMatchIn(stringLiteral)) {
                try {
                    val userInfo = URI(stringLiteral).userInfo
                    if (userInfo != null) {
                        val parts = userInfo.split(":").toTypedArray()
                        return (parts.size > 1 && parts[0] != parts[1]) && !(parts.size == 2 && parts[1].isEmpty())
                    }
                } catch (e: URISyntaxException) {
                    // ignore, stringLiteral is not a valid URI
                }
            }
            return false
        }

        private fun isQuery(value: String, match: String): Boolean {
            val followingString = value.substring(value.indexOf(match) + match.length)
            return (followingString.startsWith("=?")
                || followingString.startsWith("=%")
                || followingString.startsWith("=:")
                || followingString.startsWith("={") // string format
                || followingString == "='")
        }
    }

    override fun visitBinaryExpression(expression: KtBinaryExpression, context: KotlinFileContext) {
        if (expression.operationToken == KtTokens.EQ || expression.operationToken == KtTokens.PLUSEQ) {
            val left = expression.left
            left?.identifier()?.let { checkVariable(context, left, it, expression.right!!) }
        }
    }

    override fun visitProperty(property: KtProperty, context: KotlinFileContext) {
        property.initializer?.let {
            checkVariable(context, property.nameIdentifier!!, property.name!!, it)
        }
    }

    override fun visitStringTemplateExpression(expression: KtStringTemplateExpression, context: KotlinFileContext) {
        val content = if (!expression.hasInterpolation()) expression.asConstant() else ""
        if (isURIWithCredentials(content)) {
            context.reportIssue(expression, "Review this hard-coded URL, which may contain a credential.")
        } else {
            literalPatterns()
                .mapNotNull { regex -> regex.find(content) }
                .map { matchResult -> matchResult.groups[1]!!.value }
                .filter { match: String -> !isQuery(content, match) }
                .forEach { credential: String -> context.report(expression, credential) }
        }
    }

    private fun KtElement.isNotEmptyString() = this is KtStringTemplateExpression
        && !this.hasInterpolation()
        && this.asConstant().isNotEmpty()

    private fun KotlinFileContext.report(tree: PsiElement, matchName: String) {
        reportIssue(tree, "\"$matchName\" detected here, make sure this is not a hard-coded credential.")
    }

    private fun KotlinFileContext.checkAssignedValue(matchResult: MatchResult, regex: Regex, leftHand: PsiElement, value: String) {
        if (!regex.containsMatchIn(value)) {
            report(leftHand, matchResult.groups[1]!!.value)
        }
    }

    private fun KtExpression.identifier(): String? = when (this) {
        is KtNameReferenceExpression -> getReferencedName()
        is KtDotQualifiedExpression -> selectorExpression?.identifier()
        else -> null
    }

    private fun checkVariable(ctx: KotlinFileContext, variable: PsiElement, variableName: String, value: KtElement) {
        if (value.isNotEmptyString()) {
            variablePatterns()
                .mapNotNull { regex -> regex.find(variableName)?.let { it to regex } }
                .forEach { (matcher, regex) ->
                    ctx.checkAssignedValue(
                        matcher,
                        regex,
                        variable,
                        (value as KtStringTemplateExpression).asConstant())
                }
        }
    }

    private fun variablePatterns() = variablePatterns ?: toPatterns("")

    private fun literalPatterns() = literalPatterns ?: toPatterns("""=\S""")

    private fun toPatterns(suffix: String): Sequence<Regex> {
        return credentialWords.split(",").toTypedArray()
            .asSequence()
            .map { obj: String -> obj.trim { it <= ' ' } }
            .map { word: String -> Regex("($word)$suffix", RegexOption.IGNORE_CASE) }
    }
}
